line_profiler/line_profiler.py

```diff
@@ -132,23 +164,163 @@ def print_stats(self, stream=None, output_unit=None, stripzeros=False,
                   stream=stream, stripzeros=stripzeros,
                   details=details, summarize=summarize, sort=sort, rich=rich)
 
-    def add_module(self, mod):
-        """ Add all the functions in a module and its classes.
+    def _add_namespace(self, duplicate_tracker, namespace, *,
+                       match_scope='none', wrap=False):
+        count = 0
+        add_cls = functools.partial(self._add_namespace, duplicate_tracker,
+                                    match_scope=match_scope, wrap=wrap)
+        add_func = self.add_callable
+        remember = duplicate_tracker.add
+        wrap_func = self.wrap_callable
+        wrap_failures = {}
+        if match_scope == 'none':
+            def check(*_):
+                return True
+        elif isinstance(namespace, type):
+            check = self._add_class_filter(namespace, match_scope)
+        else:
+            check = self._add_module_filter(namespace, match_scope)
+
+        for attr, value in vars(namespace).items():
+            if id(value) in duplicate_tracker:
+                continue
+            remember(id(value))
+            if isinstance(value, type):
+                if check(value) and add_cls(value):
+                    count += 1
+                continue
+            try:
+                if not add_func(value):
+                    continue
+            except TypeError:  # Not a callable (wrapper)
+                continue
+            if wrap:
+                wrapper = wrap_func(value)
+                if wrapper is not value:
+                    try:
+                        setattr(namespace, attr, wrapper)
+                    except (TypeError, AttributeError):
+                        # Corner case in case if a class/module don't
+                        # allow setting attributes (could e.g. happen
+                        # with some builtin/extension classes, but their
+                        # method should be in C anyway, so
+                        # `.add_callable()` should've returned 0 and we
+                        # shouldn't be here)
+                        wrap_failures[attr] = value
+            count += 1
+        if wrap_failures:
+            msg = (f'cannot wrap {len(wrap_failures)} attribute(s) of '
+                   f'{namespace!r} (`{{attr: value}}`): {wrap_failures!r}')
+            warnings.warn(msg, stacklevel=2)
+        return count
+
+    @staticmethod
+    def _add_module_filter(mod, match_scope):
```

Comment:
Feels weird  to have these as staticmethods. Maybe if we change the match_scope to a StrEnum, we can add these as methods there?